File: medium/3. Longest Substring Without Repeating Characters/Solution.java
class Solution {
   public int lengthOfLongestSubstring(String s) {
        HashMap<Character, Integer> map = new HashMap<>();
        int maxLength = 0;
        int left = 0;

        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            if (map.containsKey(c)) {
                left = Math.max(left, map.get(c) + 1);
            }
            map.put(c, i);
            maxLength = Math.max(maxLength, i - left + 1);
        }

        return maxLength;
    }
}

File: medium/3. Longest Substring Without Repeating Characters/Solution.kt
class Solution {
    fun lengthOfLongestSubstring(s: String): Int {
        val map = HashMap<Char, Int>()
        var maxLength = 0
        var left = 0
        
        for(i in s.indices) {
            val c = s[i]
            
            if(map.containsKey(c)) {
                left = left.coerceAtLeast(map[c]!! + 1)
            }
            map[c] = i
            maxLength = maxLength.coerceAtLeast(i - left + 1)
        }
        
        return maxLength
    }
}